Add ref for cumsum

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86229

As noted in the comment, this decomposition may not be as efficient as
specific implementations of it in different backends. Added here to then
benchmark it. Note that this is needed by TorchInductor https://github.com/pytorch/pytorch/issues/93631

lezcano committed:
Add ref for cumsum

As noted in the comment, this decomposition may not be as efficient as
specific implementations of it in different backends. Added here to then
benchmark it. Note that this is needed by TorchInductor https://github.com/pytorch/torchdynamo/issues/883

[ghstack-poisoned]

lezcano commented:
Actually, yep, it does take forever for large sizes.
```python
[-------------------- cumsum --------------------]
                           |  Decomposed  |  Eager
32 threads: --------------------------------------
      (256, 256, 256), -1  |      12      |   1.19
      (256, 256, 256), -2  |      13      |   1.00
      (256, 256, 512), -1  |      27      |   2.38
      (256, 512, 256), -2  |      55      |   2.03
      (512, 512, 256), -1  |      31      |   4.75
      (512, 256, 512), -2  |      54      |   3.80
      (512, 512, 512), -1  |     169      |   9.49
      (512, 512, 512), -2  |     220      |   7.69

Times are in milliseconds (ms).
```

What should we do with this one @ngimel @mruberry?

ngimel commented:
Yeah it's expected to take forever. What does jax do? I suspect as a generic decomposition this has little utility, because backend will have to have backend-specific lowering. This approach could be useful for some non-standard reductions that don't have `cum*` counterpart, where the alternative is sequential reduction.

lezcano commented:
They have two algorithms. On CUDA, when the reduction is less than 32, they do something that I assume it's similar to what we do here. If the reduction is larger, they have a divide-and-conquer algorithm implemented in `associative_scan`. See https://jax.readthedocs.io/en/latest/_modules/jax/_src/lax/control_flow/loops.html#associative_scan This one launches several kernels.

lezcano committed:
Update on "Add ref for cumsum"


As noted in the comment, this decomposition may not be as efficient as
specific implementations of it in different backends. Added here to then
benchmark it. Note that this is needed by TorchInductor https://github.com/pytorch/torchdynamo/issues/883

[ghstack-poisoned]

lezcano commented:
I could implement an `associative_scan` of sorts, but the main issue is that it uses `concatenate`, so it may not be super efficient either...

lezcano commented:
@pytorchbot merge
